### ISSUE TITLE
Memory leak in ManualResetEventSlim

### ISSUE THREAD

**Issue by ayushm007** [mono/mono#19665]
## Steps to Reproduce

1. Install Mono version 6.6.0.161
2. Run attached program
Note: program run on ARM target on linux.

### Current Behavior

Below chart showing the memory taken by program. It is continuously increase and looks like it hampered GC cycle also.
![ManualResetEventSlim](https://user-images.githubusercontent.com/8023411/80448505-23cd7f80-893a-11ea-9e3e-953d28a82fd3.png)

Download the source code from here
[ManualResetEventSlim.zip](https://github.com/mono/mono/files/4543416/ManualResetEventSlim.zip)

### Expected Behavior

Memory should be used with in a range. Same program with ManualResetEvent uses memory in below pattern.
![ManualResetEvent](https://user-images.githubusercontent.com/8023411/80449269-4d87a600-893c-11ea-8118-381edeb7ddcc.png)

Download the source code from here
[ManualResetEvent.zip](https://github.com/mono/mono/files/4543417/ManualResetEvent.zip)

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:

Mono JIT compiler version 6.6.0.161 (tarball Thu Feb 27 16:07:49 IST 2020)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       normal
        Notifications: epoll
        Architecture:  armel,vfp+hard
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        Suspend:       preemptive
        GC:            sgen (concurrent by default)

## Stacktrace

```
Please paste the stack trace here if available.
```

**Comment by ayushm007** [comment]
Any suggestions about this issue?

**Comment by BrzVlad** [comment]
I can reproduce increasing high memory usage with this, which comes from managed memory, on random amd64 platform. I'm not sure if this is problem from testcase or bcl or somewhere else. I'll investigate in a couple of days.

**Comment by ayushm007** [comment]
Hi BrzVlad, 
Thanks for update. I just want to share that we found leak also in ManualResetEvent program when run it for a long period i.e. 4 to 5 days.

Also, I would like to know how I can do further investigation. Any tool which you suggest for finding the leak on linux platform.

**Comment by BrzVlad** [comment]
@ayushm007 I'm not aware of great tools to use on linux. Normally I would recommend doing snapshots with Xamarin.Profiler (macos) and see what kind of objects are contributing to the leak.

Having said that, this looks like it's caused by a leak in enumerating a concurrent queue. I can no longer reproduce this, by switching to the new netcore profile, so I assume it was fixed in the meantime in the bcl.

**Comment by ayushm007** [comment]
@BrzVlad Is it fixed in mono bcl? Can you please share me the link from where i can get this fix
.

**Comment by BrzVlad** [comment]
This is actually not really a bug or a leak, but rather an extreme case of conservative stack scanning leading to memory waste due to the way the concurrent queue is implemented. For me, the memory will eventually go down, but sometime only after hundreds of MBs are being consumed. Currently, you could workaround this either by avoiding the use of concurrent queue or using coreclr which precisely scans the stack (.netcore 3.1).

@stephentoub `ConcurrentQueue` is implemented using linked `ConcurrentQueueSegment`. To my understanding, even when a segment is cleared, all the instances are still linked together. So if a segment reference is getting stuck somewhere on the stack, the concurrent queue will forever leak. Could we make this queue more conservative gc friendly by breaking the segment chain in some slowpath ?

**Comment by stephentoub** [comment]
> Could we make this queue more conservative gc friendly by breaking the segment chain in some slowpath ?

Probably, but I expect it's not trivial.  We rely heavily on the GC doing the "right thing" in order to properly clean up here.  The issue is around anything that enumerates the contents of the queue.  If it were just about enqueue and dequeue, then the thread dequeueing the last element of a segment could also unlink the segment.  But with enumeration (e.g. GetEnumerator, CopyTo, ToArray, etc.) and snapshot semantics, when the last element of a segment is dequeued, that element and that segment may still be needed for any in-flight enumerations: if we were to erroneously null out the next segment link, we could break such enumerations, causing them to miss a bunch of elements they should have actually seen.  So addressing the GC issue by breaking these segments would require more active tracking of in-flight enumerations, with some kind of reference counting on each segment to indicate how many enumerators still need that segment, and only nulling out the next link when no more active enumerators are present nor ever will be.  I'm sure it's doable, but I don't know how much overhead it would add nor whether it would be pay-for-play.

**Comment by BrzVlad** [comment]
I remember seeing locks around enumeration/snapshot. Maybe we could safely break it off there ? Even if it doesn't guarantee us anything if the queue is will not be enumerated on, it might still prevent the vast majority of leaks, including in the scenario from this issue, while not requiring any redesign or adding overhead.

**Comment by stephentoub** [comment]
> I remember seeing locks around enumeration/snapshot.

Yes, but the current scheme simply marks each segment that it's required, and once required, that's never undone, i.e. there's no reference counting, or needing to deal with the possibility of a race between one enumeration ending and another enumeration starting or a segment being dequeued.

> Even if it doesn't guarantee us anything if the queue is will not be enumerated on, it might still prevent the vast majority of leaks, including in the scenario from this issue, while not requiring any redesign or adding overhead.

I'm more concerned that we may inadvertently break enumeration in some cases.

Please feel free to try prototyping something.

**Comment by ayushm007** [comment]
@BrzVlad Thanks for suggestion. 
Now it looks fine, after avoiding ConcurrentQueue with below custom code.  
`	

	public class ConcurrentQueueLite<T>
	{
		
		private List<T> mQueue;
		private int mListCounter;

		public ConcurrentQueueLite ()
		{
			mQueue = new List<T> ();
			mListCounter = 0;
		}
		public void Enqueue(T item)
		{
			lock (mQueue) {
				mQueue.Add (item);
				mListCounter++;
			}
		}
		public bool TryDequeue(out T item)
		{
			item = default(T);
			lock (mQueue) {
				if (mListCounter > 0) {
					item = mQueue [0];
					mQueue.RemoveAt (0);
					mListCounter--;
				}
			}
			return item != null;
		}
		public int Count
		{
			get {return mListCounter;}
		}
		public T FirstOrDefault (Func<T, bool> predicate)
		{
			if (predicate == null)
			{
				throw new Exception("condition missing");
			}
			lock (mQueue) {
				foreach (T local in mQueue) {
					if (predicate (local)) {
						return local;
					}
				}
				return default(T);
			}
		}
	}
`

![Snap](https://user-images.githubusercontent.com/8023411/82111037-a25e5580-975f-11ea-91ac-cb16dd58a324.JPG)

Is there any improvement or fix regarding this in any other mono version? So that i can test with that version also.

**Comment by BrzVlad** [comment]
There is no available fix yet and not sure when it will come, on mono. As suggested above, your other alternative would be to migrate your project over to .net core, using a completely different runtime.

**Comment by nealef** [comment]
We've been looking at this issue for a client who was experiencing large and growing memory use where it appeared lots of `QueueSegement` objects were resident in the heap.  Early experiments showed that nulling of the `Tail.Next` field had a positive effect but was not sufficient. There are two reasons:
1. If debugging is enabled then it is possible that the threadpool QueueSegments will be enumerated in which case a race condition may exist. This may be addressed by the use of a Spinlock.
2. It is possible for a thread to hold the value of the current tail but due to thread scheduling another may update the queueTail and render the other thread’s value stale. Damien has provided a detailed explanation.
### Stale Tail Value
Thread T1 grabs an old segment which happens to be depleted (...) but is followed by a new segment full of items (III).  It then goes to sleep.
```
    T1
      \
QT -- [...] -> [III] -X
```
While it sleeps, T2 grabs the same tail, advances QT to the next segment, and dequeues its first item:
```
    T1
      \
QT -- [...] -> [III] -X
      /
    T2

    T1
      \
       [...] -> [.II] -X
               /
             QT
```
In the new code, T2 sets the Next pointer of the first segment to `null`:
```
   T1
     \
      [...] -X [.II] -X
              /
            QT
```
This causes T1 to believe it as reached the end of the queue when it wakes up.  Without the (new) step above, it would have continued and successfully dequeued an item from the second segment.
#### Test Program
We created a test program that illustrates the problem and has described in detail what the program has revealed. The test program, would examine the "race" and "leak" situations.
##### Race Condition
The test program was able to highlight race conditions between the dequeuing and enumeration operations. For example:
* We have two threads which grab the same version of queueTail in Dequeue, and fail to tail.TryDequeue
* The first one manages to update queueTail, and to advance to the next segment after null’ing tail.Next
* The second one, which happens to be lagging a bit, compares its (stale) tail.Next to null and believes it has reached the end of the queue, that there is no work left.
This is not a crash, and the subsystem is likely to recover, but it still is an abnormal exit.
##### Leak Condition
That program makes it very easy to reproduce the leak.  It will, based on user input, generate QueueSegments and force garbage collection.
Executing it under the log profiler is sufficient to see held QueueSegments. e.g.:
```
┌────
│ mono --profile=log:alloc,heapshot=1gc threadpool-bits.exe leak
│ 
│ mprof-report --verbose --traces output.mlpd
└────
```
Here is a fragment of heapshot output which shows that none of the 10 QueueSegments which have been allocated have been released even though the queue has advanced past them: it corresponds to the complete chain and shows two mysterious root references:
```
┌────
│   400         10       40 Test.ThreadPoolWorkQueue.QueueSegment
│    2 root references (0 pinning)
│    9 references from: Test.ThreadPoolWorkQueue.QueueSegment
│    2 references from: Test.ThreadPoolWorkQueue
└────
```
Digging with GDB, I could find the first QueueSegment which had been allocated on the stack of thread #12, 36 words above the current stack pointer.  It naturally holds the rest of the chain.
```
┌────
│   240         10       24 Test.ThreadPoolWorkQueue.QueueSegment (bytes: +0, count: +0)
│    6 root references (0 pinning)
│    9 references from: Test.ThreadPoolWorkQueue.QueueSegment
│    2 references from: Test.ThreadPoolWorkQueue
└────
```
(In one of the runs, I could find the pointer on five thread stacks without much effort.)
Here is what happens if one ‘null’s the ‘Next’ pointers (I ignored the race conditions for this test):
```
┌────
│   80          2       40 Test.ThreadPoolWorkQueue.QueueSegment (bytes: +0, count: +0)
│    2 root references (0 pinning)
│    2 references from: Test.ThreadPoolWorkQueue
└────
```
Unsurprisingly, the root references are still there—but only two segments are held: the one stuck on the stack, and the latest one.
Success!  (Well, if one figures out a way of safely nulling these pointers, that is…)
#### Remediation
The object of the remediation is to allow the mono garbage collector to dispose of the unused `QueueSegments` and avoid any race condition should the enumerator be active.

The aim here is to protect grabbing and enumerating through the chain of QueueSegments but without the overheads of traditional locking or semaphores. The proposal is that the probability of performing the debugging tasks that would invoke this enumeration performance is not an important factor so using a `SpinLock()` that would consume CPU while in operation may be an acceptable method.

The proposed fix we developed changes the enumerator so it doesn't grab the `queueTail` until much later so it may be protected by a `SpinLock` that holds for the minimal amount of time. Secondly, we don't process one element at a time but, instead, do what the `ConcurrentQueue` system does and create a copy of the current chain configuration. We lock on the creation of the copy but then release and then iterate through the copy. In this way the debugger which is iterating through this chain can spend as long as it likes on each element.

Meanwhile the `Dequeue` operation can safely null the `Tail.Next` field by protecting with the `SpinLock`.

[threadpool-null.patch.txt](https://github.com/mono/mono/files/6737762/threadpool-null.patch.txt)

**Comment by gkarabin** [comment]
This is an interesting thread. I ran into it while observing similar behavior in a Xamarin.Android application that use BlockingCollection in a few places. Does anyone know or suspect that it would have similar issues hanging onto memory? We often see hundreds of MB accumulating, then a big drop after hours, on devices with sufficient free RAM. On other devices we see the system run into OOM first.

**Comment by BrzVlad** [comment]
@gkarabin Sounds plausible. `BlockingCollection` uses by default a `ConcurrentQueue` as a container internally

**Comment by gkarabin** [comment]
@BrzVlad - thanks! I'll take a closer look at `ConcurrentQueueSegment` in the profiler. I don't recall seeing them in significant numbers before but it's a thread to look for in the noise.

**Comment by danports** [comment]
Are there any plans to revisit this issue? @BrzVlad I see you tried to write a fix earlier, but that didn't pan out? I've got a similar problem on .NET Core on Linux, where the application gets OOM killed after consuming a few GB of memory apparently due to `ConcurrentQueue` holding on to references to large object trees even when the queue is actually empty. The repro project and analysis is at https://github.com/getsentry/sentry-dotnet/issues/2516 since I originally thought this was an issue with the Sentry .NET SDK, but it appears now that Sentry's only problem is that it uses `ConcurrentQueue` internally. 😅

**Comment by danports** [comment]
The memory consumption issue I have with `ConcurrentQueue` has been consistently reproducible on .NET 7 on both Linux and Windows. `ConcurrentQueue` appears to maintain references to elements that have since been removed from the queue, even when the queue is empty. So perhaps there is a separate issue with `ConcurrentQueue` unrelated to the GC differences between Mono and CoreCLR? I just jumped on this issue since the symptoms seemed similar to what I was experiencing.

**Comment by stephentoub** [comment]
> ConcurrentQueue appears to maintain references to elements that have since been removed from the queue, even when the queue is empty.

Can you share a repro?

Are you enumerating / peeking at the queue in addition to dequeuing from it?

**Comment by jamescrosswell** [comment]
> Can you share a repro?

@danports [originally provided this](https://github.com/getsentry/sentry-dotnet/issues/2516#issuecomment-1654689536), which I've been able to use to successfully reproduce the issue targeting `net7.0`. 

I've also confirmed that there is no memory leak if, instead of using `ConcurrentDictionary<T>`, Sentry's BackgroundWorker uses [some alternative](https://github.com/getsentry/sentry-dotnet/issues/2516#issuecomment-2099585468). 

The repro unfortunately has a certain degree of complexity to it. It needs a connection to send events to Sentry and it also needs a valid MySQL Connection string. If you want, I'm happy to provide a valid Sentry DSN and a docker-compose file to fire up an instance of mariadb that will work with Dan's sample. 

> Are you enumerating / peeking at the queue in addition to dequeuing from it?

Yes, Sentry's BackgroundWorker [Peeks](https://github.com/getsentry/sentry-dotnet/blob/ee3b6c60ed1c9b38f527eed44dce6cb3a4b05e8e/src/Sentry/Internal/BackgroundWorker.cs#L144) an item from the queue... does some stuff with it and, if successful, then [dequeue's](https://github.com/getsentry/sentry-dotnet/blob/ee3b6c60ed1c9b38f527eed44dce6cb3a4b05e8e/src/Sentry/Internal/BackgroundWorker.cs#L179) the item (discarding the result).
